BitwardenShared/Core/Platform/Services/AuthenticatorSyncService.swift

```diff
@@ -0,0 +1,247 @@
+        sharedKeychainRepository: SharedKeychainRepository,
+        stateService: StateService,
+        vaultTimeoutService: VaultTimeoutService
+    ) {
+        self.authBridgeItemService = authBridgeItemService
+        self.cipherService = cipherService
+        self.clientService = clientService
+        self.configService = configService
+        self.errorReporter = errorReporter
+        self.sharedKeychainRepository = sharedKeychainRepository
+        self.notificationCenterService = notificationCenterService
+        self.stateService = stateService
+        self.vaultTimeoutService = vaultTimeoutService
+        super.init()
+    }
+
+    // MARK: Public Methods
+
+    public func start() {
+        Task {
+            if await configService.getFeatureFlag(FeatureFlag.enableAuthenticatorSync,
+                                                  defaultValue: false) {
+                subscribeToAppState()
+            }
+        }
+    }
+
+    // MARK: Private Methods
+
+    /// Check to see if the shared Authenticator key exists. If it doesn't already exist, create it.
+    ///
+    private func createAuthenticatorKeyIfNeeded() async throws {
+        let storedKey = try? await sharedKeychainRepository.getAuthenticatorKey()
+        if storedKey == nil {
+            let key = SymmetricKey(size: .bits256)
+            let data = key.withUnsafeBytes { Data(Array($0)) }
+            try await sharedKeychainRepository.setAuthenticatorKey(data)
+        }
+    }
+
+    /// Take a list of encrypted ciphers, decrypt them, filter for only active ciphers with a totp code, then
+    /// convert the list to AuthenticatorSyncItemDataModel to be stored and sync'd to the Authenticator app.
+    ///
+    /// - Parameters:
+    ///   - ciphers: The encrypted `Cipher` objects.
+    ///   - userId: The userId of the account to which these Ciphers belong.
+    ///
+    /// - Returns: The decrypted, filtered, and sorted `CipherDTO` objects.
+    ///
+    private func decryptTOTPs(_ ciphers: [Cipher],
+                              userId: String) async throws -> [AuthenticatorBridgeItemDataView] {
+        let decryptedCiphers = try await ciphers.asyncMap { cipher in
+            try await self.clientService.vault().ciphers().decrypt(cipher: cipher)
+        }
+        let account = try await stateService.getActiveAccount()
+        let username = account.profile.name ?? account.profile.email
+
+        let ciphersToUse = decryptedCiphers.filter { cipher in
+            cipher.deletedDate == nil
+                && cipher.type == .login
+                && cipher.login?.totp != nil
+        }
+
+        return ciphersToUse.map { cipher in
+            AuthenticatorBridgeItemDataView(
+                favorite: false,
+                id: cipher.id ?? UUID().uuidString,
+                name: cipher.name,
+                totpKey: cipher.login?.totp,
+                username: username
+            )
+        }
```

Comment:
🎨 One can do the filtering before decrypting the ciphers to improve performance as we don't want all ciphers decrypted but only the ones who comply with the filter. This reduces the number of calls to the SDK and also the waiting time on decrypting unneeded ciphers.
Moreover, we can use the filter before decrypting because we don't need any decrypted value to perform the filter correctly:
```suggestion
    private func decryptTOTPs(_ ciphers: [Cipher],
                              userId: String) async throws -> [AuthenticatorBridgeItemDataView] {
        let totpCiphers = ciphers.filter { cipher in
            cipher.deletedDate == nil
                && cipher.type == .login
                && cipher.login?.totp != nil
        }
        let decryptedCiphers = try await totpCiphers.asyncMap { cipher in
            try await self.clientService.vault().ciphers().decrypt(cipher: cipher)
        }
        let account = try await stateService.getActiveAccount()
        let username = account.profile.name ?? account.profile.email

        return decryptedCiphers.map { cipher in
            AuthenticatorBridgeItemDataView(
                favorite: false,
                id: cipher.id ?? UUID().uuidString,
                name: cipher.name,
                totpKey: cipher.login?.totp,
                username: username
            )
        }
```

Reply:
👍 Good note. Fixed in latest.